kornia/contrib/sam/architecture/sam.py

```diff
@@ -0,0 +1,134 @@
+            prompt_encoder: Encodes various types of input prompts.
+            mask_decoder: Predicts masks from the image embeddings and encoded prompts.
+            pixel_mean: Mean values for normalizing pixels in the input image.
+            pixel_std: Std values for normalizing pixels in the input image.
+            device: The desired device to be used in the model
+        """
+        super().__init__()
+        self.image_encoder = image_encoder
+        self.prompt_encoder = prompt_encoder
+        self.mask_decoder = mask_decoder
+        self.register_buffer("pixel_mean", tensor(pixel_mean, device=device).view(-1, 1, 1), False)
+        self.register_buffer("pixel_std", tensor(pixel_std, device=device).view(-1, 1, 1), False)
+        self._device = device
+
+    @property
+    def device(self) -> Device:
+        return self._device
+
+    @no_grad()
+    def forward(self, batched_input: list[dict[str, Any]], multimask_output: bool) -> list[dict[str, Tensor]]:
+        """Predicts masks end-to-end from provided images and prompts.
+
+        If prompts are not known in advance, using SamPredictor is recommended over calling the model directly.
+
+        Args:
+            batched_input: A list over input images, each a dictionary with the following keys. A prompt key can be
+                excluded if it is not present. The options are:
+
+                - "image": The image as a torch tensor in 3xHxW format, already transformed for input to the model.
+                - "original_size": (tuple(int, int)) The original size of the image before transformation, as (H, W).
+                - "point_coords": (Tensor) Batched point prompts for this image, with shape BxNx2. Already
+                  transformed to the input frame of the model.
+                - "point_labels": (Tensor) Batched labels for point prompts, with shape BxN.
+                - "boxes": (Tensor) Batched box inputs, with shape Bx4. Already transformed to the input frame of the
+                  model.
+                - "mask_inputs": (Tensor) Batched mask inputs to the model, in the form Bx1xHxW.
+
+            multimask_output: Whether the model should predict multiple disambiguating masks, or return a single mask.
+
+        Returns:
+            A list over input images, where each element is as dictionary with the following keys.
+                - "masks": (Tensor) Batched binary mask predictions, with shape BxCxHxW, where B is the number
+                  of input prompts, C is determiend by multimask_output, and (H, W) is the original size of the image.
+                - "iou_predictions": (Tensor) The model's predictions of mask quality, in shape BxC.
+                - "low_res_logits": (Tensor) Low resolution logits with shape BxCxHxW, where H=W=256. Can be passed as
+                  mask input to subsequent iterations of prediction.
+        """
+        input_images = stack([self.preprocess(x["image"]) for x in batched_input], dim=0)
```

Comment:
Kornia normalize + pad ? But I actually I’d remove this preprocess step because is not very complaint if you want compose with other Kornia ops. I’d assume that the input to SAM is already a normalized image

Reply:
Plus with our ops can be performed all vectorised

Reply:
I think this and the post-process should be part of the SamPredictor

Reply:
I’d the image comes from a pre Kornia process, it will come as [0,1] which means that needs to be de normalised. Since we are making this model part of our ecosystem I’d make it complaint to the rest of the library, assume that the input is normalized. Preprocessing is a requirement yes, but is evil to be added as part of the model. Can lead to a lots of bugs.  That’s possibly something we should cover as part of the Image api, to track whether an image is normalized

Reply:
opinions here @shijianjian @ducha-aiki

Reply:
check the changes I did on https://github.com/kornia/kornia/pull/2315/commits/22b1d66f7b3beebdf8f5da22c29b409a673d7d78 

with this, the Sam arch code (`sam.architecture.sam.Sam(...)`) is cleaner with just the init and forward, then a class (`sam.model.Sam(...)`) with the others utilities stuff (from pre-trained, build, etc). Also, this forward should be a little lighter because is expected for the batched images to be given and didn't compute the upscaling.

if this organization makes sense for you too, in a subsequent commit, the sam prediction should be more aligned with the official one, since now we can simply use the model call to predict.